contracts/optimism/TokenRateOracle.sol

```diff
@@ -0,0 +1,210 @@
+import {IChainlinkAggregatorInterface} from "./interfaces/IChainlinkAggregatorInterface.sol";
+import {CrossDomainEnabled} from "./CrossDomainEnabled.sol";
+import {Initializable} from "@openzeppelin/contracts/proxy/utils/Initializable.sol";
+import {Versioned} from "../utils/Versioned.sol";
+
+interface ITokenRateOracle is ITokenRateUpdatable, IChainlinkAggregatorInterface {}
+
+/// @author kovalgek
+/// @notice Oracle for storing token rate.
+/// @dev Token rate updates can be delivered from two sources: L1 token rate pusher and L2 bridge.
+contract TokenRateOracle is CrossDomainEnabled, ITokenRateOracle, Versioned, Initializable {
+
+    /// @dev Stores the dynamic data of the oracle. Allows safely use of this
+    ///     contract with upgradable proxies
+    struct TokenRateData {
+        /// @notice wstETH/stETH token rate.
+        uint192 tokenRate;
+        /// @notice L1 time when token rate was pushed.
+        uint64 rateL1Timestamp;
+    } // occupy a single slot
+
+    /// @dev Location of the slot with TokenRateData
+    bytes32 private constant TOKEN_RATE_DATA_SLOT =
+        keccak256("TokenRateOracle.TOKEN_RATE_DATA_SLOT");
+
+    /// @notice A bridge which can update oracle.
+    address public immutable L2_ERC20_TOKEN_BRIDGE;
+
+    /// @notice An address of account on L1 that can update token rate.
+    address public immutable L1_TOKEN_RATE_PUSHER;
+
+    /// @notice A time period when token rate can be considered outdated.
+    uint256 public immutable TOKEN_RATE_OUTDATED_DELAY;
+
+    /// @notice A time period when token rate can be considered outdated.
+    uint256 public constant MAX_ALLOWED_L1_L2_TIME_DIFFERENCE = 86400;
+
+    /// @notice Number of seconds in one day.
+    uint256 public constant ONE_DAY_SECONDS = 86400;
+
+    /// @notice Allowed token rate deviation per day in basic points.
+    uint256 public constant TOKEN_RATE_DEVIATION = 500;
+
+    /// @notice Decimals of the oracle response.
+    uint8 public constant DECIMALS = 18;
+
+    /// @notice Basic point scale.
+    uint256 private constant BASIS_POINT_SCALE = 1e4;
+
+    /// @param messenger_ L2 messenger address being used for cross-chain communications
+    /// @param l2ERC20TokenBridge_ the bridge address that has a right to updates oracle.
+    /// @param l1TokenRatePusher_ An address of account on L1 that can update token rate.
+    /// @param tokenRateOutdatedDelay_ time period when token rate can be considered outdated.
+    constructor(
+        address messenger_,
+        address l2ERC20TokenBridge_,
+        address l1TokenRatePusher_,
+        uint256 tokenRateOutdatedDelay_
+    ) CrossDomainEnabled(messenger_) {
+        L2_ERC20_TOKEN_BRIDGE = l2ERC20TokenBridge_;
+        L1_TOKEN_RATE_PUSHER = l1TokenRatePusher_;
+        TOKEN_RATE_OUTDATED_DELAY = tokenRateOutdatedDelay_;
+
+        _disableInitializers();
+    }
+
+    function initialize(uint256 tokenRate_, uint256 rateL1Timestamp_) external initializer {
```

Comment:
should think twice about the upgrade voting (how and when should the initialization call happen)